### PR TITLE
Support both slowpath and fastpath models (9)

### PR DESCRIPTION
**Pull request by sulakshm**
Dynamically choose which Io path to choose based on device add parameters (enable_fp).
And rework congestion logic.

### CODE REVIEW

File: dev.c [hunk 1]
```diff
@@ -607,52 +642,88 @@ static int copy_in_read_data_iovec(struct iov_iter *iter,
 	return 0;
 }
 
-static int fuse_notify_read_data(struct fuse_conn *conn, unsigned int size,
-				struct iov_iter *iter)
+static int __fuse_notify_read_data_slowpath(struct fuse_conn *conn,
```
Comment (prabirpaul) [inline dev.c]: Lot of duplication between the functions below. Can you instead pass in fastpath as argument to a common fn or reorg however you prefer so common code gets moved out?

File: pxd.c [hunk 2]
```diff
@@ -254,6 +252,8 @@ static struct fuse_req *pxd_fuse_req(struct pxd_device *pxd_dev, int nr_pages)
 		printk_ratelimited(KERN_ERR "%s: request alloc (%d pages) failed: %d",
 			 __func__, nr_pages, status);
 	}
+
+	if (req) req->fastpath = pxd_dev->fastpath;
```
Comment (prabirpaul) [inline pxd.c]: req can't be null here.

File: pxd_core.h [hunk 3]
```diff
@@ -36,6 +39,7 @@ struct pxd_device {
 	struct pxd_context *ctx;
 	bool connected;
 	mode_t mode;
+	bool fastpath;
```
Comment (prabirpaul) [inline pxd_core.h]: Why is fastpath per device property? Are we planning to allow mixed set of devices?

File: pxd.c [hunk 4]
```diff
@@ -596,22 +596,27 @@ static int pxd_init_disk(struct pxd_device *pxd_dev, struct pxd_add_out *add)
...
 
-	/* cannot choose io processing model dynamically based on queue size.
-	 * Because other files also need to know how requests are to be processed.
-	 */
-#if !defined(USE_REQUESTQ_MODEL)
-	q = blk_alloc_queue(GFP_KERNEL);
-	if (!q) {
-		err = -ENOMEM;
-		goto out_disk;
+#ifndef __PX_FASTPATH__
```
Comment (prabirpaul) [inline pxd.c]: This is not needed.

File: pxd_fastpath.h [hunk 5]
```diff
@@ -71,6 +71,10 @@ struct pxd_fastpath_extension {
 	char device_path[MAX_PXD_BACKING_DEVS][MAX_PXD_DEVPATH_LEN+1];
 
 	struct thread_context *tc;
+#define DEFAULT_CONGESTION_THRESHOLD (256)
+#define MAX_CONGESTION_THRESHOLD (1024)
```
Comment (prabirpaul) [inline pxd_fastpath.h]: This can't be greater than PXD_MAX_QDEPTH.

File: Makefile.in [hunk 6]
```diff
@@ -20,27 +20,79 @@ endif
...
 ifeq ($(shell test "$(KVERSION)" = "4.18.0-80.el8.x86_64"  -o  "$(KVERSION)" = "4.18.0-80.1.2.el8.x86_64" -o "$(KVERSION)" = "4.18.0.el8.x86_64"; echo $$?),0)
 PXDEFINES += -D__PX_BLKMQ__
-$(warning enabling slow path with BLKMQ support)
 endif
 
+## fastpath specific checks
+ifeq ($(call verlater,${kmajor},${fp_major}),0)
+PXDEFINES += -D__PX_FASTPATH__
+px-objs += pxd_fastpath.o
+$(warning kernel fast path enabled, current kernel version $(KVERSION) need minimum $(FPATH_MINKVER))
```
[the remark this answers is not in the document]
Comment (sulakshm) [reply Makefile.in]: It is compatible, those are logs to understand from build process to know whether fastpath got compiled in.

File: pxd_core.h (hunk 3 above)
Comment (sulakshm) [reply pxd_core.h]: Ofcourse it is per device. I believe we had a discussion on this yesterday. Devices shall be attached in fastpath only upon request from px and still can always dynamically switch to slowpath if there are any kind of IO errors as a fallback.

**Comment by sulakshm** [comment]
On then general comment about cleaning up duplicate code, I shall work on it and push again. But apart from that, please do continue to see whether changes are ready.

File: pxd_fastpath.h (hunk 5 above)
Comment (sulakshm) [reply pxd_fastpath.h]: revised, default chooses to  use PXD_MAX_QDEPTH, but through sysfs can override if needed upto MAX_CONGESTION_THRESHOLD. There is no default q depth definition.

File: pxd.c (hunk 2 above)
Comment (sulakshm) [reply pxd.c]: revised

File: pxd.c (hunk 4 above)
Comment (sulakshm) [reply pxd.c]: Why? This is fastpath requests on unsupported kernel.

**Pull request merged**